File: typescript/src/security/Ed25519Service.ts
/**
 * Ed25519 signing service using TweetNaCl
 * SPDX-License-Identifier: MIT
 */

import nacl from "tweetnacl";

export interface Ed25519KeyPair {
  privateKey: Uint8Array; // 32-byte seed
  publicKey: Uint8Array; // 32-byte public key
}

/**
 * Ed25519 signing service using TweetNaCl/libsodium
 * Key format: 32-byte seed (private), 32-byte point (public), 64-byte signature
 */
export class Ed25519Service {
  /**
   * Generate a new Ed25519 key pair
   * @returns {Ed25519KeyPair} Tuple of (privateKey, publicKey)
   */
  static generateKeyPair(): Ed25519KeyPair {
    const keyPair = nacl.sign.keyPair();
    return {
      privateKey: keyPair.secretKey.slice(0, 32),
      publicKey: keyPair.publicKey,
    };
  }

  /**
   * Sign data using an Ed25519 private key
   * @param privateKey - 32-byte Ed25519 seed
   * @param data - The data to sign
   * @returns 64-byte Ed25519 signature
   */
  static sign(privateKey: Uint8Array, data: Uint8Array): Uint8Array {
    if (privateKey.length !== 32) {
      throw new Error("Ed25519 private key must be 32 bytes");
    }

    // TweetNaCl expects a 64-byte secret key (32 seed + 32 public key)
    // We reconstruct it from the seed
    const keyPair = nacl.sign.keyPair.fromSeed(privateKey);
    const signed = nacl.sign(data, keyPair.secretKey);

    // Extract just the signature (first 64 bytes)
    return signed.slice(0, 64);
  }

  /**
   * Verify an Ed25519 signature
   * @param publicKey - 32-byte Ed25519 public key
   * @param data - The signed data
   * @param signature - 64-byte Ed25519 signature
   * @returns true if the signature is valid
   */
  static verify(
    publicKey: Uint8Array,
    data: Uint8Array,
    signature: Uint8Array
  ): boolean {
    if (publicKey.length !== 32) {
      return false;
    }

    if (signature.length !== 64) {
      return false;
    }

    try {
      // Combine signature + data for TweetNaCl verification
      const signedMessage = new Uint8Array(64 + data.length);
      signedMessage.set(signature, 0);
      signedMessage.set(data, 64);

      const opened = nacl.sign.open(signedMessage, publicKey);
      return opened !== null;
    } catch {
      return false;
    }
  }

  /**
   * Verify a signature with fallback support for legacy P-256 keys
   * Currently only supports Ed25519 (P-256 fallback not implemented in TS version)
   * @param publicKey - Public key bytes (32 = Ed25519)
   * @param data - The signed data
   * @param signature - The signature bytes
   * @returns true if the signature is valid
   */
  static verifyWithFallback(
    publicKey: Uint8Array,
    data: Uint8Array,
    signature: Uint8Array
  ): boolean {
    // Standard Ed25519 path
    if (publicKey.length === 32) {
      return this.verify(publicKey, data, signature);
    }

    // Legacy P-256 path would go here, but not implemented in TS version
    return false;
  }
}

File: python/aether/security/ed25519_service.py
# SPDX-License-Identifier: MIT

"""Ed25519 signing service using PyNaCl (libsodium)."""

import nacl.signing
import nacl.exceptions
from typing import Tuple


class Ed25519SigningService:
    """
    Static Ed25519 signing service using PyNaCl/libsodium.

    Key format:
    - Private key: 32-byte seed
    - Public key: 32-byte point
    - Signature: 64-byte signature
    """

    @staticmethod
    def generate_keypair() -> Tuple[bytes, bytes]:
        """
        Generate a new Ed25519 key pair.

        Returns:
            A tuple of (private_key: 32 bytes, public_key: 32 bytes).
        """
        signing_key = nacl.signing.SigningKey.generate()
        private_key = bytes(signing_key)
        public_key = bytes(signing_key.verify_key)
        return private_key, public_key

    @staticmethod
    def sign(private_key: bytes, data: bytes) -> bytes:
        """
        Sign data using an Ed25519 private key.

        Args:
            private_key: 32-byte Ed25519 seed.
            data: The data to sign.

        Returns:
            64-byte Ed25519 signature.

        Raises:
            ValueError: If private_key is not 32 bytes.
        """
        if private_key is None:
            raise ValueError("private_key cannot be None")
        if data is None:
            raise ValueError("data cannot be None")

        if len(private_key) != 32:
            raise ValueError(
                f"Ed25519 private key must be 32 bytes, got {len(private_key)}"
            )

        signing_key = nacl.signing.SigningKey(private_key)
        signed_message = signing_key.sign(data)
        # Extract just the signature (first 64 bytes of the signed message)
        return signed_message.signature

    @staticmethod
    def verify(public_key: bytes, data: bytes, signature: bytes) -> bool:
        """
        Verify an Ed25519 signature.

        Args:
            public_key: 32-byte Ed25519 public key.
            data: The signed data.
            signature: 64-byte Ed25519 signature.

        Returns:
            True if the signature is valid, False otherwise.
        """
        if public_key is None or data is None or signature is None:
            return False

        if len(public_key) != 32:
            return False

        if len(signature) != 64:
            return False

        try:
            verify_key = nacl.signing.VerifyKey(public_key)
            verify_key.verify(data, signature)
            return True
        except (nacl.exceptions.BadSignatureError, ValueError):
            return False

    @staticmethod
    def verify_with_fallback(
        public_key: bytes, data: bytes, signature: bytes
    ) -> bool:
        """
        Verify a signature with fallback support for legacy keys.

        For now, this only supports Ed25519. Future versions may add P-256 ECDSA
        fallback during migration windows.

        Args:
            public_key: Public key bytes (32 = Ed25519).
            data: The signed data.
            signature: The signature bytes.

        Returns:
            True if the signature is valid.
        """
        return Ed25519SigningService.verify(public_key, data, signature)
